src/mesh/amr_loadbalance.cpp

```diff
@@ -462,368 +697,183 @@ void Mesh::RedistributeAndRefineMeshBlocks(ParameterInput *pin, ApplicationInput
-              received[rb_idx] = true;
-            }
+
+        if (oloc.level == nloc.level &&
+            ranklist[on] != Globals::my_rank) { // same level, different rank
+#ifdef MPI_PARALLEL
+          for (auto &var : pb->vars_cc_) {
+            if (!finished[idx])
+              finished[idx] =
+                  TryRecvSameToSame(n - nbs, ranklist[on], var.get(), pb.get(), this);
+            all_received = finished[idx++] && all_received;
           }
-          rb_idx++;
+#endif
         } else if (oloc.level > nloc.level) { // f2c
           for (int l = 0; l < nleaf; l++) {
-            if (ranklist[on + l] == Globals::my_rank) continue;
-            if (!received[rb_idx]) {
-              PARTHENON_MPI_CHECK(
-                  MPI_Test(&(req_recv[rb_idx]), &test, MPI_STATUS_IGNORE));
-              if (static_cast<bool>(test)) {
-                FinishRecvFineToCoarseAMR(pb.get(), recvbuf[rb_idx], loclist[on + l]);
-                received[rb_idx] = true;
+            auto pob = pb;
+            if (ranklist[on + l] == Globals::my_rank) pob = old_block_list[on + l - onbs];
+            LogicalLocation &oloc = loclist[on + l];
+            for (auto &var : pb->vars_cc_) {
+              if (!finished[idx]) {
+                auto var_in = pob->meshblock_data.Get()->GetVarPtr(var->label());
+                finished[idx] =
+                    TryRecvFineToCoarse(n - nbs, ranklist[on + l], oloc, var_in.get(),
+                                        var.get(), pb.get(), this);
               }
+              all_received = finished[idx++] && all_received;
             }
-            rb_idx++;
           }
-        } else { // c2f
-          if (ranklist[on] == Globals::my_rank) continue;
-          if (!received[rb_idx]) {
-            PARTHENON_MPI_CHECK(MPI_Test(&(req_recv[rb_idx]), &test, MPI_STATUS_IGNORE));
-            if (static_cast<bool>(test)) {
-              FinishRecvCoarseToFineAMR(pb.get(), recvbuf[rb_idx]);
-              received[rb_idx] = true;
+        } else if (oloc.level < nloc.level) { // c2f
+          for (auto &var : pb->vars_cc_) {
+            if (!finished[idx]) {
+              auto pob = pb;
+              if (ranklist[on] == Globals::my_rank) pob = old_block_list[on - onbs];
+              auto var_in = pob->meshblock_data.Get()->GetVarPtr(var->label());
+              finished[idx] = TryRecvCoarseToFine(
+                  n - nbs, ranklist[on], nloc, var_in.get(), var.get(), pb.get(), this);
             }
+            all_received = finished[idx++] && all_received;
           }
-          rb_idx++;
         }
       }
       // rb_idx is a running index, so we repeat the loop until all vals are true
-    } while (!std::all_of(received.begin(), received.begin() + rb_idx,
-                          [](bool v) { return v; }));
-    Kokkos::fence();
+    } while (!all_received && niter < 1e7);
+    if (!all_received) PARTHENON_FAIL("AMR Receive failed");
```

Comment:
No need to fence here any more?

Reply:
Isn't this fence required in the old code because `recvbuf` is used in the set kernels but is then deleted below?

Reply:
IIRC we added this because (under the assumption we'd allow for multiple device execution spaces. e.g., cuda streams) all data transfers should be done prior to additional operations.
I don't think that a single fence here would hurt performance, but it might be a safer choice (in the long run) given that  `ProlongateCellCenteredValues` is called right afterwards.
What do you think? (and feel free to merge).

Reply:
I agree that it is probably not a performance bottleneck and may save us from issues down the road. I will add a fence before prolongation.